feat: add market value and sell possibility for Possession

Phylosius opened the pull request:
- Added the possibility of selling Possession (only for IMMOBILISATION and ENTREPRISE)
- Added PatriLang grammar for market value addition and Possession selling (only for IMMOBILISATION and ENTREPRISE)

<!-- Thanks to all the contributors; -->
<!-- the 3th Group of PROJET1 of HEI J -->

Phylosius committed:
chore: specify vendable

Phylosius committed:
chore: add the target account in Vendable::vendre

Phylosius committed:
chore: create sell information class

Johnathon47 committed:
fix: update version of lombok in current version

Johnathon47 committed:
feat: change the version to orginal

Johnathon47 committed:
Merge branch 'main' of https://github.com/Ny-Antsaniaina/patrimoine into jobanana

Phylosius committed:
feat: add ValeurMarche in Vendable on creation

Phylosius committed:
refactor: use Set for ValeurMarche list

Phylosius committed:
chore: add fees addition to the compteBeneficiaire in sell confirmation

Phylosius committed:
feat: add sell logic to Possession

Johnathon47 committed:
Merge branch 'main' of https://github.com/Ny-Antsaniaina/patrimoine into jobanana

Phylosius committed:
style: format

Johnathon47 committed:
feat: add condition to change valeurComptable to 0 after the 'DateDeVente'

Johnathon47 committed:
Merge branch 'jobanana'

Ny-Antsaniaina committed:
feat: add type ENTREPRISE in TypeArgent

NESWILLY committed:
feat: add ENTREPRISE type in TypeAgregrat

NESWILLY committed:
feat: add the logic for ValeurMarche with each TypeAgregat.

NESWILLY committed:
feat: implement getvaleurMarches in Possession class

Ny-Antsaniaina committed:
Merge branch 'main' into ben

Ny-Antsaniaina committed:
fix: edit getValeurMarche

Ny-Antsaniaina committed:
style: format

Johnathon47 committed:
feat: include DateDeVente in the condition to format valeurComptable from the DateDeVente and after

Phylosius committed:
refactor: rename ValeurMarche into ValeurMarchee

Phylosius committed:
feat: create class for possession sell

Phylosius committed:
chore: made InformationDeVente serializable

Phylosius committed:
style: format

Phylosius committed:
chore: add null check for sale date in get comptable value

Phylosius committed:
fix: verification on possession's sale value getting

Phylosius committed:
chore: add getter to sell class

Ny-Antsaniaina committed:
feat: add class VenteTest

Ny-Antsaniaina committed:
chore: test vente on class VenteTest

Phylosius committed:
test: test market value creation

Phylosius committed:
chore: add sell and market value operation grammar for PatriLang

Phylosius committed:
chore: specify sell market visitor for PatriLang

Phylosius committed:
chore: specify sell market visitor for PatriLang

Phylosius committed:
chore: made Vente extends Possession

Phylosius committed:
chore: made Vente extends Possession

Phylosius committed:
style: format

Phylosius committed:
style: format

NESWILLY committed:
feat: add checking type of Possession in addValeurMarche and vendre methods

Phylosius committed:
refactor: made ajoutValeurMarchee and ventePossession like existing grammar in PatriLang

Phylosius committed:
refactor: made ajoutValeurMarchee and ventePossession like existing grammar in PatriLang

NESWILLY committed:
fix: change type checking of Possession in addValeurMarche and vendre methods

Ny-Antsaniaina committed:
Merge branch 'main' into ben

Phylosius committed:
chore: correct Vente::projectionFuture

Phylosius committed:
feat: create sell visitors for PatriLang

Ny-Antsaniaina committed:
Merge remote-tracking branch 'origin/duduk' into ben

Ny-Antsaniaina committed:
Merge branch 'duduk' into ben

NESWILLY committed:
Merge remote-tracking branch 'origin/duduk' into neswilly

Ny-Antsaniaina committed:
feat : add VaeurMarcheVisitor

NESWILLY committed:
feat: merge the valeurMarcheeVisitor

Johnathon47 committed:
feat: change assertEquals to assertThrows for type non valid for the test

Johnathon47 committed:
feat: update test

Phylosius committed:
feat: create sell visitor for PatriLang

Phylosius committed:
feat : add ValeurMarcheVisitor

Phylosius committed:
update

Phylosius committed:
fix: possession interpretation in market value visitor

Phylosius committed:
fix: possession to sell interpretation in ValeurMarcheeVisitor

NESWILLY committed:
feat(test): add test for ValeurMarcheeVisitor to throws exception when possession type is not supported.

NESWILLY committed:
feat(test): add test for VenteVisitor to throws exception when possession type is not supported.

NESWILLY committed:
test: add venteVisitor and valeurMarcheeVisitor tests

NESWILLY committed:
fix(test): add the scope to ValeurMarcheeVisitor and VenteVisitor

Phylosius committed:
feat: add possession translator in variable visitor

Phylosius committed:
feat: use VariableVisitor::asPossession in VenteVisitor and ValeurMarcheeVisitor

Phylosius committed:
chore: update visitors test

Phylosius committed:
refactor: remove unused code

Phylosius committed:
refactor: use VariableVisitor::asPossession in ValeurMarcheeVisitor and VenteVisitor for PatriLang

Phylosius committed:
refactor: remove Vendable::getObjectifs because of non-usage

Phylosius committed:
chore: remove Vendable::getObjectifs because of non-usage

Phylosius committed:
style: google reformating

RickaPrincy commented on src/main/java/school/hei/patrimoine/modele/possession/Possession.java:
```diff
@@ -51,6 +58,12 @@ public final Devise devise() {
   public final Argent valeurComptableFuture(LocalDate tFutur) {
+    if (getDateDeVente() != null) {
+      if (!tFutur.isBefore(getDateDeVente())) {
+        return valeurComptable.mult(0);
```
C1. valeurComtpableFuture should'nt be touched ... the calc should be done on getValeurComptable()

Something like t.isBefore(getDateDeVente())  for example

And return new Argent(0, devise) instead of multiplying by 0

RickaPrincy commented on src/main/java/school/hei/patrimoine/modele/possession/Possession.java:
```diff
@@ -67,4 +80,53 @@ public String nom() {
+
+  @Override
+  public void vendre(Argent valeurDeVente, LocalDate dateDeVente, Compte compteBeneficiaire) {
+    if ((typeAgregat() != TypeAgregat.IMMOBILISATION) && (typeAgregat() != TypeAgregat.ENTREPRISE))
```
C2. No , we should be able to sell everything

RickaPrincy commented on src/main/java/school/hei/patrimoine/modele/possession/Vente.java:
```diff
@@ -0,0 +1,35 @@
+import school.hei.patrimoine.modele.Argent;
+
+@Getter
+public final class Vente extends Possession {
```
C3. Finally, let's say that Vente doesn't inherit from Possession.

RickaPrincy commented on src/main/java/school/hei/patrimoine/modele/possession/Vente.java:
```diff
@@ -0,0 +1,35 @@
+
+  @Override
+  public Vente projectionFuture(LocalDate tFutur) {
+    return new Vente(tFutur, getPossessionAVendre(), Argent.ariary(0));
```
import static for ariary(0)

RickaPrincy commented on src/main/java/school/hei/patrimoine/modele/vente/InformationDeVente.java:
```diff
@@ -0,0 +1,60 @@
+        .filter(v -> v.t().isBefore(t) || v.t().isEqual(t))
+        .sorted(Comparator.comparing(ValeurMarchee::t))
+        .findFirst()
+        .orElse(new ValeurMarchee(t, Argent.euro(0)));
```
import static

RickaPrincy commented on src/main/java/school/hei/patrimoine/modele/vente/InformationDeVente.java:
```diff
@@ -0,0 +1,60 @@
+    return valeurDeVente != null && dateDeVente != null && compteBeneficiaire != null;
+  }
+
+  public void confirmeVente(
```
Confirmer Vente d'une information de Vente doesn't really have a sense.

RickaPrincy commented on src/main/java/school/hei/patrimoine/modele/vente/Vendable.java:
```diff
@@ -0,0 +1,24 @@
+
+  public Argent getValeurDeVente();
+
+  public LocalDate getDateDeVente();
```
I mean, maybe it's better if it's return an optional instead of null

RickaPrincy commented on src/main/java/school/hei/patrimoine/patrilang/grammar/PatriLangParser.g4:
```diff
@@ -92,6 +92,15 @@ objectif
     :   MUL id COMMA? dateValue=variable COMMA? MOT_CORRIGER valeurComptable=variable MOT_DANS compteNom=variable
     ;
+
+ajoutValeurMarchee
```
valeurMarchee

RickaPrincy commented on src/main/java/school/hei/patrimoine/patrilang/grammar/PatriLangParser.g4:
```diff
@@ -92,6 +92,15 @@ objectif
+    :   MUL id COMMA? dateDeVente=variable COMMA? MOT_VALEUR_MARCHEE_DE prixDeVente=variable MOT_POUR possessionAffectee=variable
+    ;
+
+ventePossession
```
vente

RickaPrincy commented on src/test/java/school/hei/patrimoine/modele/vente/ValeurMarcheeTest.java:
```diff
@@ -0,0 +1,40 @@
+            dateValeurMarchee,
+            valeurPcInitiale.minus(Argent.ariary(20_000), dateValeurMarchee));
+
+    assertEquals(pc.getValeurMarche(dateValeurMarchee.plusDays(1)), subject);
```
expected on the left (always), actual on the right

RickaPrincy commented on src/test/java/school/hei/patrimoine/patrilang/unit/visitors/possession/ValeurMarcheeVisitorTest.java:
```diff
@@ -0,0 +1,46 @@
+
+    var exception =
+        assertThrows(
+            UnsupportedOperationException.class,
```
IllegalException

RickaPrincy commented on src/test/java/school/hei/patrimoine/patrilang/unit/visitors/possession/VenteVisitorTest.java:
```diff
@@ -0,0 +1,46 @@
+  UnitTestVisitor visitor =
+      new UnitTestVisitor() {
+        @Override
+        public Object visitVentePossession(PatriLangParser.VentePossessionContext ctx) {
```
public Vente visitVentePossession...

RickaPrincy commented:
CI is failing.

Phylosius committed:
refactor: rename operation ajoutValeurMarchee to valeurMarchee for PatriLang

Phylosius committed:
refactor: rename operation venteOperation to vente for PatriLang

Phylosius committed:
refactor: update sell operation name for PatriLang

Johnathon47 committed:
fix: valeurComptableFutur on possession return Argent(0, devise()) and delete the condition of vente to not accept to sell everything just two kinds of "agrégat".

Phylosius committed:
refactor: use dynamic device resolution for getValueMarche

Phylosius committed:
refactor: use dynamic device resolution in Vente::projectionFuture

Phylosius committed:
refactor: use dynamic devise resolution when computing Argent value

Phylosius committed:
refactor(test): use explicit type on VenteVisitorTest

Used Vente instead of Object.

Phylosius committed:
refactor(test): use explicit type on ValeurMarcheeVisitorTest

Used ValeurMarchee instead of Object.

Phylosius committed:
refactor(test): correct type casting in PatriLang's visitors tests

Phylosius committed:
update

Phylosius committed:
feat\!: made all possession sellable and put sell logic in Possession::vendre instead of in InformationDeVente::confirmeVente

Cause it makes more sense

Phylosius committed:
feat: made all possession types sellable and put sell logic in Possession

Johnathon47 committed:
Merge branch 'main' of https://github.com/Ny-Antsaniaina/patrimoine into jobanana

Johnathon47 committed:
refactor: rm condition who change valeurComptable on method projectionFutur

Johnathon47 committed:
feat: add condition to change valeurComptable to 0 after the 'DateDeVente' t is include and if datevente is not null too

Johnathon47 committed:
refactor: valeurComptableFutur and add condition in ValeurComptable

Johnathon47 committed:
refactor: use format code

Ny-Antsaniaina committed:
feat: add materiel type t patrilang scope

Johnathon47 committed:
fix: prevent NullPointerException in getValeurMarche

Because the new ValeurMarchee in devise take the getValeurDeVente (return null by default) while he will take devise of getValeurMarche. So I create method a getDeviseProche to avoid the StackOverflow if a use this getValeurMarchee(t).valeur.devise().

Johnathon47 committed:
fix/nullpointer-getValeurMarche

Phylosius committed:
feat: add remover for VariableVisitor

Phylosius committed:
test: materiel addition to patrilang scope

Phylosius committed:
chore: remove unused test ressource

Phylosius committed:
style: format

Phylosius committed:
fix: device rnot resolved in InformationDeVente::getValeurMarchee

Phylosius committed:
feat: affected materiel now added to patrilang variable scope when performing possedeMateriel operation

Phylosius committed:
fix: outated test for VenteVisitor

Phylosius committed:
fix: Possession::valeurComptable return null

Johnathon47 committed:
fix: I add informationVente in projectionFutur for Each Possession for save the information of sell to fix null informationDeVente in projectionFutur.

Johnathon47 committed:
refactor: use format code

Johnathon47 committed:
fix: Null informationDeVente in ProjectionFutur

Phylosius committed:
chore(test): remove VariableVisitor::removeFromScope and use new variable visitor instead
